### PR TITLE
add Statistics for cache miss rate

### PR DESCRIPTION
**Pull request by luohaha**
For a better user experience, add optional statistics for cache miss rate.

### CODE REVIEW

**Comment by p0mvn** [comment]
Hi @luohaha 

Thank you for this change. This is certainly useful to keep track of.

I have one question though - how do you expect this new statistic to be read or queried?

**Comment by luohaha** [comment]
@p0mvn After new mutable tree, we can get a statistic with it. We can periodic read this statistic and set to a new metric, and then report to prometheus or print to log.

**Comment by p0mvn** [comment]
Since `nodeDb` is an unexported field of the mutable tree, how can we read the statistic periodically? Is it possible to expose a method for doing so that would also synchronize access to the statistics?

File: options.go [hunk 1]
```diff
@@ -10,6 +29,9 @@ type Options struct {
 	// this, an error is returned when loading the tree. Only used for the initial SaveVersion()
 	// call.
 	InitialVersion uint64
+
+	// When Stat is not nil, statistical logic needs to be executed
+	Stat *Statistics
```
Comment (p0mvn) [inline options.go:34]: Is it possible to define this on `nodeDb` instead?

Options are used for configuring iavl. On the other hand, this seems non-configuration specific

Comment (luohaha) [reply options.go:34]: https://github.com/facebook/rocksdb/blob/main/include/rocksdb/options.h#L609
Usage is just like rocksdb, eg:
```
Options options;
options.statistics = ROCKSDB_NAMESPACE::CreateDBStatistics();
Status s = DB::Open(options, kDBPath, &db);
...
options.statistics->getTickerCount(NUMBER_BLOCK_COMPRESSED);
```
Our usage:
```
opt.Statistics = MyStatistics
NewMutableTreeWithOpts(db, 100, &opt);
...
print(MyStatistics)
```

Comment (p0mvn) [reply options.go:34]: Thanks for the example.

Although it might be convenient in C++, I don't think it would be as pretty in Go

Relying on `opts.Statistics` pointer is prone to data races. In the linked example, the synchronization of the counters is likely handled in `getTickerCount(...)`

In our case, we currently don't have such a method. However, even if we had one, relying on the user on calling that method instead of accessing `opt.Statistics` directly is dangerous. Someone could easily misuse it and access one of the counters directly without a lock (e.g. `opts.Statistics.CacheHitCnt`) causing a data race.

So I think we could do one of the following:
- avoid exporting `CacheHitCnt` and `CacheMissCnt` and create methods like:
```
func (s *Statistics) GetCacheHitCnt() {...}
```
in which we atomically get the counters

OR

- define the Statistics on `nodeDb` instead of `Options`, expose getters on the mutable tree and atomically get the counters

Please let me know what you think

Comment (luohaha) [reply options.go:34]: I prefer expose getters on the Statistics than mutable tree, because more statistics item may be added in the future, and their getters will be more concise to be hidden on the Statistics.
And I also prefer to add Statistics to option, because statistical behavior could cause performance damage and no every one can accept this. So it is better to be optional.

Comment (p0mvn) [reply options.go:34]: Good call, sounds good to me

File: nodedb.go [hunk 2]
```diff
@@ -125,11 +125,18 @@ func (ndb *nodeDB) GetNode(hash []byte) (*Node, error) {
 
 	// Check the cache.
 	if elem, ok := ndb.nodeCache[string(hash)]; ok {
+		if ndb.opts.Stat != nil {
```
Comment (p0mvn) [inline nodedb.go:128]: Could you please also add a couple of unit tests with and without the `nil` `Stat`. 

Also, asserting that the counters get increased would be great.

File: options.go [hunk 3]
```diff
@@ -1,5 +1,37 @@
 package iavl
 
+import "sync/atomic"
+
+// Statisc about db runtime state
+type Statistics struct {
+	// Each time GetNode and GetFastNode operation hit cache
+	cacheHitCnt uint64
```
Comment (p0mvn) [inline options.go:8]: We mainly have 2 types of nodes:
1. Regular (that you get with `GetNode(...)`
2. Fast (`GetFastNode(...)`

They each have their own cache (`nodeCache`, `fastNodeCache`):
https://github.com/cosmos/iavl/blob/640ad78af7f34486f4793b3e442c8d20c751daf0/nodedb.go#L79-L83

What do you think about having a separate counter for each cache?

File: nodedb.go (hunk 2 above)
Comment (luohaha) [reply nodedb.go:128]: Done

File: options.go (hunk 3 above)
Comment (luohaha) [reply options.go:8]: That is great. Done

File: tree_test.go [hunk 4]
```diff
@@ -1982,3 +1982,69 @@ func Benchmark_GetByIndex(b *testing.B) {
 		}
 	})
 }
+
+func TestNodeCacheStatisic(b *testing.T) {
```
Comment (p0mvn) [inline tree_test.go:1986]: nit: suggest changing to `t` since it is idiomatic to name benchmarks as `b` but regular tests as `t`

```suggestion
func TestNodeCacheStatisic(t *testing.T) {
```

File: tree_test.go [hunk 5]
```diff
@@ -1982,3 +1982,69 @@ func Benchmark_GetByIndex(b *testing.B) {
...
+	db, err := db.NewDB("test", db.MemDBBackend, "")
+	require.NoError(b, err)
+
+	// with cache enable
+	b.Run("with_cache", func(sub *testing.T) {
```
Comment (p0mvn) [inline tree_test.go:1991]: Seems like a good candidate to make this table-driven to avoid repetition. Do you mind making it so?

File: tree_test.go [hunk 6]
```diff
@@ -1982,3 +1982,69 @@ func Benchmark_GetByIndex(b *testing.B) {
...
+			key := []byte(strconv.Itoa(i))
+			val, err := it.Get(key)
+			require.NoError(b, err)
+			require.NotNil(b, val)
+			require.NotEmpty(b, val)
+		}
+
+		require.Equal(b, 0, int(opts.Stat.GetFastCacheHitCnt()))
+		require.Equal(b, numKeyVals, int(opts.Stat.GetFastCacheMissCnt()))
+		require.Equal(b, 1, int(opts.Stat.GetCacheMissCnt()))
```
Comment (p0mvn) [inline tree_test.go:2048]: Can we make an assertion on `GetCacheHitCnt` too, please?

File: tree_test.go [hunk 7]
```diff
@@ -1982,3 +1982,69 @@ func Benchmark_GetByIndex(b *testing.B) {
...
+		for i := 0; i < numKeyVals; i++ {
+			key := []byte(strconv.Itoa(i))
+			val, err := it.Get(key)
+			require.NoError(b, err)
+			require.NotNil(b, val)
+			require.NotEmpty(b, val)
+		}
+		require.Equal(b, numKeyVals, int(opts.Stat.GetFastCacheHitCnt()))
+		require.Equal(b, 0, int(opts.Stat.GetFastCacheMissCnt()))
+		require.Equal(b, 1, int(opts.Stat.GetCacheHitCnt()))
```
Comment (p0mvn) [inline tree_test.go:2017]: Let's make an assertion on miss as well?

**Review by p0mvn — COMMENTED** [review]
Great work so far, a couple more comments. Let me know what you think

File: tree_test.go (hunk 5 above)
Comment (luohaha) [reply tree_test.go]: Done.

Comment (p0mvn) [reply tree_test.go]: Table-driven tests look like the following:
```
func TestSplit(t *testing.T) {
    tests := map[string]struct {
        input string
        sep   string
        expected  []string
    }{
        "simple":       {input: "a/b/c", sep: "/", expected: []string{"a", "b", "c"}},
        "wrong sep":    {input: "a/b/c", sep: ",", expected: []string{"a/b/c"}},
        "no sep":       {input: "abc", sep: "/", expected: []string{"abc"}},
        "trailing sep": {input: "a/b/c/", sep: "/", expected: []string{"a", "b", "c"}},
    }

    for name, tc := range tests {
        t.Run(name, func(t *testing.T) {
            got := Split(tc.input, tc.sep)
            if !reflect.DeepEqual(tc.expected, got) {
                t.Fatalf("expected: %v, got: %v", tc.want, got)
            }
        })
    }
}
```

where there is a `tests` slice and then you iterate over each and run it. Each test case has a name, inputs, and expected values.

Source: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

A good example in this repo: https://github.com/cosmos/iavl/blob/abaaacdcf919da1fb9fb780862346dcb99b60fd4/internal/encoding/encoding_test.go#L11

File: tree_test.go [hunk 8]
```diff
@@ -1982,3 +1982,49 @@ func Benchmark_GetByIndex(b *testing.B) {
...
 	})
 }
+
+func TestNodeCacheStatisic(t *testing.T) {
+	db, err := db.NewDB("test", db.MemDBBackend, "")
+	require.NoError(t, err)
+	const numKeyVals = 100000
+	testcases := map[string]struct {
+		cacheSize int
+		expect    []int
```
Comment (p0mvn) [inline tree_test.go:1992]: Instead of having this as `[]int`, can we split it into a field for each.
E.g.
```
expectFastHit int
expectFastMiss int
...
```

Currently, it's not intuitive which value is represented by which index in `expect`

**Review by p0mvn — APPROVED** [review]
LGTM once the final comment is addressed. Great work

File: tree_test.go (hunk 8 above)
Comment (luohaha) [reply tree_test.go]: Done.

**Comment by marbar3778** [comment]
Thanks for the reviews @p0mvn

**Comment by luohaha** [comment]
@p0mvn @marbar3778 When will we publish IAVL-v0.18.1 ? After this, I can add some metrics about cache miss rate in cosmos-sdk rely on IAVL-v0.18.1

**Comment by marbar3778** [comment]
it is part of our current sprint, should happen over the next two weeks
